modules/message-lane/src/lib.rs

```diff
@@ -90,19 +159,237 @@ decl_module! {
 
+		/// Block initialization.
+		fn on_initialize(now: T::BlockNumber) -> Weight {
+			let mut dispatcher = T::QueuedMessageDispatch::with_any_weight();
+			let (queued_messages_dispatched, weight_spent_on_queued_dispatch) = dispatcher.dispatch();
+			if queued_messages_dispatched != 0 {
+				frame_support::debug::trace!(
+					target: "runtime",
+					"Spent {} weight on {} queued messages from on_initialize()",
+					weight_spent_on_queued_dispatch,
+					queued_messages_dispatched,
+				);
+			}
+
+			weight_spent_on_queued_dispatch
+		}
+
 		/// Send message over lane.
 		#[weight = 0] // TODO: update me (https://github.com/paritytech/parity-bridges-common/issues/78)
 		pub fn send_message(
 			origin,
 			lane_id: LaneId,
 			payload: T::Payload,
-		) {
-			let _ = ensure_signed(origin)?;
+			delivery_and_dispatch_fee: T::MessageFee,
+		) -> DispatchResult {
+			let submitter = ensure_signed(origin)?;
+
+			// let's first check if message can be delivered to target chain
+			T::TargetHeaderChain::verify_message(&payload).map_err(|err| {
+				frame_support::debug::trace!(
+					target: "runtime",
+					"Message to lane {:?} is rejected by target chain: {:?}",
+					lane_id,
+					err,
+				);
+
+				Error::<T, I>::MessageRejectedByChainVerifier
+			})?;
+
+			// now let's enforce any additional lane rules
+			T::LaneMessageVerifier::verify_message(
+				&submitter,
+				&delivery_and_dispatch_fee,
+				&lane_id,
+				&payload,
+			).map_err(|err| {
+				frame_support::debug::trace!(
+					target: "runtime",
+					"Message to lane {:?} is rejected by lane verifier: {:?}",
+					lane_id,
+					err,
+				);
+
+				Error::<T, I>::MessageRejectedByLaneVerifier
+			})?;
+
+			// let's withdraw delivery and dispatch fee from submitter
+			T::MessageDeliveryAndDispatchPayment::pay_delivery_and_dispatch_fee(
+				&submitter,
+				&delivery_and_dispatch_fee,
+			).map_err(|err| {
+				frame_support::debug::trace!(
+					target: "runtime",
+					"Message to lane {:?} is rejected because submitter {:?} is unable to pay fee {:?}: {:?}",
+					lane_id,
+					submitter,
+					delivery_and_dispatch_fee,
+					err,
+				);
+
+				Error::<T, I>::FailedToWithdrawMessageFee
+			})?;
+
+			// finally, save message in outbound storage and emit event
 			let mut lane = outbound_lane::<T, I>(lane_id);
-			let nonce = lane.send_message(payload);
+			let nonce = lane.send_message(MessageData {
+				payload,
+				fee: delivery_and_dispatch_fee,
+			});
 			lane.prune_messages(T::MaxMessagesToPruneAtOnce::get());
 
+			frame_support::debug::trace!(
+				target: "runtime",
+				"Accepted message {} to lane {:?}",
+				nonce,
+				lane_id,
+			);
+
 			Self::deposit_event(Event::MessageAccepted(lane_id, nonce));
+
+			Ok(())
+		}
+
+		/// Receive messages proof from bridged chain.
+		#[weight = DELIVERY_BASE_WEIGHT + max_dispatch_weight]
+		pub fn receive_messages_proof(origin, proof: MessagesProofOf<T, I>, max_dispatch_weight: Weight) -> DispatchResult {
```

Comment:
I feel this should rather be called `receive_and_dispatch_source_messages(origin, proof, max_weight)`.

Reply:
Why do we let the relayer specify `max_dispatch_weight`? What prevents them from setting `0` to pay the lowest fee and then rely on `on_initialize` to actually do the heavy work?

EDIT: Ah okay, we still charge the relayer for all the messages, but still I wonder if we shouldn't restrict the `minimal_max_dispatch_weight`, to prevent excessive queueing.

Reply:
Please see description - having some lower-bounds on `max_dispatch_weight` would block other transactions from being included into block. Having upper bounds would block some alternatively-incentivised relayers from submitting "process-my-messages-asap" transactions (i.e. transactions with empty messages vec and non-zero `max_dispatch_weight`).

I agree that excessive queueing could be a problem here. I have no other solutions right now than to completely remove queued messages. Please respond, if you think we should do that.

Reply:
> Please see description - having some lower-bounds on max_dispatch_weight would block other transactions from being included into block. 

But the point is that "other transactions" and messages should really have the same "priority". I.e. Currently we treat bridge messages in a special way and we allow them to be dispatched with highest possible priority, during `on_initialize` call. IMHO this should be fully tuneable and in most of the blockchains should only be a final necessity to do so. For instance it probably only makes sense for system-dispatches, i.e. where two blockchains communicate with each other, but definitely not for user messages. If we allow arbitrary message to be dispatched, it might be better/cheaper/faster for a `target` blockchain user to send a message over the bridge instead of simply sending a transaction to the `target` chain, cause he/she knows it is more priviledged regarding timely dispatch.

We should definitely make sure that delivering messages doesn't require the entire block weight, but having some sane minimum, and having relayers compete for these resources with other `target` blockchain users should simply be part of the design. The incentive layer should ensure that it's profitable for them.

Reply:
Irrelevant